### ISSUE TITLE
Deadlock in non-blocking communication after too many executions

### ISSUE THREAD

**Issue by gassmoeller** [open-mpi/ompi#11052]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

I confirmed the behavior for v4.0.3, v4.1.1, and 4.1.4 but the bug only appears with certain configurations (see below).

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

I checked for the bug in different configurations:
4.0.3 delivered via apt with standard Ubuntu 20.04 on a workstation: bug not present
4.0.3 and 4.1.1 installed from source by the HPC support of the affected cluster: bug is present [ompi_info.txt](https://github.com/open-mpi/ompi/files/9974104/ompi_info.txt)
4.1.4 compiled from source release by myself with minimal configuration on the same cluster: bug not present [ompi_info.txt](https://github.com/open-mpi/ompi/files/9984320/ompi_info.txt)
4.1.4 compiled from source release by myself with identical configuration to 4.1.1 and 4.0.3 above: bug is present. [ompi_info.txt](https://github.com/open-mpi/ompi/files/9984323/ompi_info.txt)

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.

N/A

### Please describe the system on which you are running

* Operating system/version: RHEL7 / 3.10.0-1062.18.1.el7.x86_64
* Computer hardware: 2 x AMD EPYC 7702 Rome
* Network type: HDR Infiniband (but bug happens when run on a single node with 2 sockets)

-----------------------------

## Details of the problem

I am observing an MPI deadlock in a non-blocking communication algorithm. I observed this in the deal.II library and posted it as a potential bug there as well (dealii/dealii#14405, @bangerth), but since I can recreate the bug entirely without deal.II and with only standard c++ and openmpi I am posting it here as well.

Description of the MPI algorithm that fails: A non-blocking some-to-some MPI communication pattern similar to https://doi.org/10.1145/1837853.1693476. Processes send requests for replies to other MPI ranks using ISend, and answer requests they receive until all requests and answers have been sent and received. Senders are unknown to receiving processes and message length is also unknown, therefore MPI_IProbe and MPI_Get_count have to be used to detect incoming messages until all processes signal that they have received all their replies by passing an MPI_IBarrier.

**Expected behavior:** The algorithm succeeds no matter how often it is executed, and the program always resumes with correctly exchanged data. This is what happens on a workstation with openmpi/4.0.3 provided by the package manager and on the affected cluster with a minimally configured openmpi/4.1.4.

**Actual behavior on the failing system:** The algorithm reliably succeeds with correctly exchanged data for around 16.7 million times before deadlocking. This happens after around 1 day in our production code, and around 5 min in the minimal example I describe below (note that the total runtime of the example is more like 30 min, because it also tests two workarounds, you can comment out the workarounds in main() for faster turnaround). The deadlock seems to happen, because all processes have passed the MPI_IBarrier, but a call to MPI_Test, checking for the result of the barrier request returns that not all processes have passed the barrier. I have verified that the issue does not appear on our local workstation on 4 and 64 MPI ranks with openmpi/4.0.3 provided by Ubuntu 20.04 via apt, but does appear on our infiniband cluster with the ompi_info listed above (it happens reliably with 128 ranks / a full node; happens most of the time with 16 ranks; and does not seem to happen with only 2 ranks). 

**Minimal example:** I have created a minimal example to reproduce the bug (I use gcc 9.3.0 and the openmpi versions listed above):
[ibarrier.cc.txt](https://github.com/open-mpi/ompi/files/9974177/ibarrier.cc.txt)
[CMakeLists.txt](https://github.com/open-mpi/ompi/files/9974180/CMakeLists.txt)

This is a simplified copy of the deal.II algorithm. The test only uses standard C++ includes and MPI, not deal.II itself. The original and much more complicated algorithm is inside: https://github.com/dealii/dealii/blob/cf72ba27c47f6932d000d5f112af9c8bcb3d1f42/include/deal.II/base/mpi_consensus_algorithms.h#L1736 (lines 1736 - 2103).

**Workarounds:** I found two workarounds for the bug in my minimal example, but both are impossible for our production code (all 3 versions, workaround 1 / 2 / and the failing example are implemented and executed in the same minimal example above).
**Workaround 1:** Because the minimal example only sends exactly 1 request and 1 reply per process with known senders/receivers I can replace the use of MPI_IProbe with MPI_Probe, which succeeds even for more than 16.7 million executions. This version calls the probe command much less often than the Iprobe in the original, but all other MPI calls should be identical except all communication should be finished before the IBarrier.
**Workaround 2:** I can duplicate the MPI_Communicator before every execution of the communication algorithm and free the duplicate afterwards. This also succeeds, but this code is performance critical for us and duplicating the communicator seems to be slow (hence the complicated communication algorithm instead of an all-to-all). I could check how much this would affect the production code or if we could implement a duplicate/free every so many executions of the algorithm, but it would be an ugly hack.

Speculations on the reasons for the bug:
- Because the algorithm reliably works for a fixed number of iterations it seems to be a resource allocation problem (unreleased and lost memory). I do not see it in the user code though, and I do not observe a significant memory leak in the program (memory use remains roughly constant, which it does not do if I for example remove a call to MPI_Wait for allocated requests; then memory explodes). Because the bug goes away if I use a freshly duplicated MPI communicator for every iteration I am wondering if there is a cache in the MPI library that is not properly released after the call to IProbe, or if too many calls to IProbe flood some internal buffer or cause a hash collision.
- Note that 16.7 million successful iterations is close to 2^24 (16777216), additional support for an internal buffer overflow or hash collision? I can find out how close we are to 2^24 if you think it is helpful.
- May this be related open-mpi/ompi#174? It has been 8 years since that ticket though, not sure if it is still relevant. But my workaround 1 will make sure all Isends are finished before running into the IBarrier, which is described as the problem in the other issue. 
- The fact that it only happens for certain configurations probably pinpoints the place in the library to search, but I know too little about the internals of openmpi to know where to look. If there are ideas for how to provoke the bug on a different system, or which configuration option to change to pinpoint the exact cause on the cluster I can do more tests.

I am pretty much out of ideas at the moment, any help or suggestions are appreciated :-).

**Comment by jsquyres** [comment]
Thank you for the detailed bug report!

1. I do notice one minor bug in your reproducer: you're calling MPI_Test on `barrier_request` inside the while loop, and only exit the loop when flag==1.  But then you call MPI_Wait on `barrier_request`.  That's technically an error -- you're waiting on a request that is already done (and should therefore be `MPI_REUQEST_NULL`).  That being said, I'm not sure why Open MPI isn't reporting that as an error to you.
2. Can you send the `configure` lines of how you built Open MPI to make the error occur?  I tried a few different ways and was unable to reproduce the error on my systems.

**Comment by jsquyres** [comment]
> Can you send the configure lines of how you built Open MPI to make the error occur? I tried a few different ways and was unable to reproduce the error on my systems.

My mistake -- I see that the `configure` lines are in the `ompi_info` output that you sent.  Let me try re-running with your configuration.

I have a few followup questions, though:

1. What version of UCX you are using?
2. Are you just invoking `mpirun -np X ./ibarrier` on a single machine to run the test? (you mentioned that the test does fail on a single machine, which definitely makes this easier to try to reproduce / debug)
3. In the case of a failed test (particularly on a single machine), are you using the UCX or the OB1 PML?

**Comment by gassmoeller** [comment]
Hi @jsquyres thanks a lot for taking the time to look into it!

To your questions:

> I do notice one minor bug in your reproducer: ...

Sorry about that. I read the following in the documentation and thought calling MPI_Wait with an already completed request is fine (https://www.open-mpi.org/doc/v3.1/man3/MPI_Wait.3.php), and just wanted to be safe:

```
One is allowed to call MPI_Wait with a null or inactive request argument. In this case the operation returns immediately with empty status. 
```

But I guess it doesnt matter either way, I removed it and the bug is not affected.

> What version of UCX you are using?

It looks like we are using UCX 1.9.0. I had to dig a bit to find this, I will confirm with my cluster support. I guess the next question will be if it happens with a newer release of UCX, I will try to find out, but it will take a bit, I have to take a week off for other projects.

> Are you just invoking mpirun -np X ./ibarrier on a single machine to run the test? (you mentioned that the test does fail on a single machine, which definitely makes this easier to try to reproduce / debug)

Yes running on a single node on the cluster with that command is enough to reproduce. It has been inconsistent with low number of ranks though (sometimes happening, sometimes not), I need to use a full node to provoke it consistently (64 or 128 ranks on two different node types on our system).

> In the case of a failed test (particularly on a single machine), are you using the UCX or the OB1 PML?

I do not know how to find out which one I am using (this is my level of experience with ompi settings :-)), but I was able to run the test prescribing to use either one with `mpirun --mca pml ucx -np 128 ...` and alternatively `mpirun --mca pml ob1 -np 128 ...`. 

It turns out using UCX consistently triggers the bug, using OB1 always works as expected. So I guess this means the bug is somewhere in UCX (or openmpi interfacing with UCX? I know too little here). It also means I have a great workaround for now, this is already quite the success. I suspect for jobs on single nodes using OB1 is ok?

**Comment by jsquyres** [comment]
MPI_REQUEST_NULL: I double checked, and you're right -- it *is* valid to call MPI_WAIT on MPI_REQUEST_NULL.  I even checked Open MPI's MPI_WAIT, and we special case handle that scenario safely.  My bad!  😳  Sorry for the noise.   Moving on...

The @open-mpi/ucx team maintains the UCX PML; I don't know much about it.  I compiled Open MPI against UCX 1.13.1 on my cluster, but when I tried to run on a single node with the UCX PML, it refuses to run (because the UCX PML does not allow itself to be selected).  I have a dim recollection that -- by default -- the UCX PML only allows itself to be run if there are IB or RoCE interfaces in the system, which I do not have.  Are you able to `mpirun --mca pml ucx -np 128 ...` *on a single system*?

Per your last paragraph where UCX *may* be at fault here, can you try upgrading to the most recent version of UCX and see if that changes anything?

FWIW: my development/testing nodes are older; I only have nodes with 24 or 32 cores.  I've been unable to get your code to fail on a single node (with ob1/vader[shared memory], since I can't get UCX to run on a single node).  I don't have IB/RoCE, so I can't use UCX across multiple nodes.

To answer your question: yes, running with ob1 on a single node should be fine.  In my case, ob1 is used on a single node automatically, but that may well be different since I have no IB/RoCE interfaces, and you assumedly do.

**Comment by jsquyres** [comment]
I just found something interesting: It looks like it's possible to reach the MPI_WAIT for `answer_request` without it being initialized (i.e., without a corresponding MPI_ISEND to fill `answer_request`).  I don't immediately see how this could happen, but I'm unfortunately out of time to stare at your example `ibarrier.cc` code to grok it deeper this morning.  Can you verify that I am correct (or not) -- that `answer_request` may be uninitialized by the time you call MPI_WAIT on it?

Calling MPI_WAIT on an uninitialized `answer_request` could definitely lead to bad behavior.

**Comment by bosilca** [comment]
I don't see how `answer_request` can be non-initialized in any setup with more than 2 processes. But for the sake or the argument you could initialize the `answer_request = MPI_REQEUST_NULL;` to always have a default value.

One major difference between the UCX PML and OB1 is the matching. UCX PML does not rely on OMPI matching but instead has its own matching, sometimes even using hardware matching. Taking in account that this code stresses the matching a lot (I really mean it, there are multiple `MPI_Iprobe` in each loop), looking at the internal UCX matching status on deadlock  could be a starting point for the @open-mpi/ucx folks.

**Comment by gassmoeller** [comment]
Thanks for all the feedback!

I wont have time to do much on this problem this week, but will get back to it next week. Feedback for now:
> Are you able to mpirun --mca pml ucx -np 128 ... on a single system?

Just to be clear the failing system is a large cluster with Infiniband interconnect, but I can provoke the problem when running on a single node of that cluster, and yes I can select ucx like you suggested run when running on a single node of that cluster. I will try next week to run it on a single system with ucx but without any IB present.

> It looks like it's possible to reach the MPI_WAIT for answer_request without it being initialized (i.e., without a corresponding MPI_ISEND to fill answer_request). I don't immediately see how this could happen ...

Yes that should be impossible. Did you do anything special to provoke this? And how did you check if answer_request was not initialized?

**Comment by jsquyres** [comment]
> Yes that should be impossible. Did you do anything special to provoke this? And how did you check if answer_request was not initialized?

Yes, sorry, let me explain more -- I wrote that up in a hurry yesterday.  

## Empirical testing

I was doing some empirical testing of the sample code.  Just for giggles, I ran it the test code through valgrind and was greatly surprised to see warnings about a conditional jump depending on an uninitialized variable inside the `MPI_Wait(&answer_request, ...)` at the bottom of the loop.  Sure enough, it actually was uninitialized -- I confirmed it two ways:

1. Exactly as George suggested, I assigned `answer_request = MPI_REQUEST_NULL` at the top of the subroutine, and the valgrind warning went away.
    * **SIDENOTE:** While this *fixes* the read-from-uninitialized problem, I do not think that it is the right *solution* -- i.e., your algorithm is designed such that it should not be possible to get to the bottom of `run()` without invoking `MPI_Isend(..., &answer_request)`.  Just making it *safe* to call `MPI_Wait(&answer_request)` does not mean that we have fixed the problem of not invoking `MPI_Isend(..., &answer_request)`, and therefore there's still something we do not understand / feels wrong.
3. I added a state variable `bool ans_initialized=false;` at the top of the subroutine, and set it to `true` immediately after both `MPI_Isend(..., &answer_request);` calls, and then added a check for `if (!ans_initialized) std::cout << "BAD!" << std::endl;` right before the `MPI_Wait(&answer_request, ...)` at the bottom of the subroutine.  This message is emitted frequently when I run with np=3 (on a single node, and therefore with ob1/vader).

I ran out of time to analyze *how* this was happening -- with my read of the code, this shouldn't have been possible -- but it does seem to be happening.  Indeed, this morning, I have re-run the tests a bunch of times, and am still getting the same results.

## Bottom line

@bosilca and I actually had to talk on the phone this morning for a good 30 minutes before we think we figured this out.  😄

We're almost certain that this is a race condition in your code based on a subtlety in MPI. 

## More detail

The subtlety is that MPI does not guarantee ordering between collective messages and point-to-point messages. 😲 

Stripping away all details: the ibarrier can actually complete before all incoming point-to-point messages have completed.  This means that the `MPI_Test(&barrier_request, ...)` may complete (correctly) *before* an `MPI_Iprobe(reply, ..., &request_is_pending, ...)` returns `request_is_pending==1`, even if there is an inbound message that would eventually satisfy the `MPI_Iprobe()`.

This may seem counter-intuitive, but consider the following simplistic explanation: MPI point-to-point messages may actually use different communication channels than MPI collective messages.  Hence, collective message B that was sent *after* point-to-point message A may actually overtake it, meaning that collective message B may complete (in the MPI sense of the word) at the target *before* point-to-point A completes at the target.

## Recommendation

It seems like what you are doing is basically an MPI_Alltoallv() operation: every process is sending a message to every other process, but each process-pair is sending a variable-length message -- and that length may actually be 0.

Your code seems to be an optimization of this scheme: don't send the zero-length messages.  Or, put differently, only send the non-zero length messages and -- at least when running at scale -- avoid a whole truckload of zero-length messages, and assumedly improve overall performance.

Indeed, in Open MPI's implementations of MPI_Alltoallv, we send zero-length messages for exactly the reasons described above: any given receiver does not know the counts from its senders, and therefore assumes that it will receive a message from *every* peer, even if the length of that message is zero.

I don't know at what scales you intend to run, or what kinds of networks you use, but have you tried using MPI_Alltoallv?  I.e., is the sending around of extra zero-length messages actually detrimental to performance?  If not, you might just want to switch to MPI_Alltoallv.  If it is detrimental to performance, you're going to need some other mechanism to know when everyone has finished sending.  This will admittedly be tricky, due to MPI's collectives and point-to-point messages not being ordered relative to each other.

**Comment by bangerth** [comment]
Hi @jsquyres -- just to confirm: Yes, the code originates from an implementation of a "consensus algorithm" that, in essence, implements a some-to-some communication where each process does not know up front who it will receive messages from. The algorithm implemented is the one from here:
```
@article{hoefler2010scalable,
  author =    {T. Hoefler and C. Siebert and A. Lumsdaine},
  title =     {Scalable communication protocols for dynamic sparse data exchange},
  journal =   {ACM Sigplan Notices},
  publisher = {ACM New York, NY, USA},
  year =      {2010},
  volume =    {45},
  number =    {5},
  pages =     {159--168},
  url =       {https://doi.org/10.1145/1837853.1693476}
}
```
The original implementation of that algorithm in deal.II was by @peterrum. I don't have access to the paper above, but I believe that his implementation followed what they described, so if our implementation has a race condition between p2p and collective operations, then so does the one described in the paper.

In typical use cases in deal.II, we may have tens of thousands of ranks, but only 5-20 messages per rank. So the "some" is really "quite sparse". My suspicion is that using `MPI_Alltoallv` would not be very efficient in that case, the main bottleneck being that *every* process has to allocate memory for *all* messages being sent, regardless of whether or not it is actually interested in that data. I believe I recall my colleague @kronbichler describing to me that that is the key reason to use these algorithms: Their memory usage is `O(1)` whereas Alltoallv will have a memory usage of `O(P)`, and so it doesn't scale even if the goal is only to find out who is going to send something to whom. (My original implementation of that last question was to use `MPI_Sum` on a vector in which every process has a 1 for each target rank it wants to send something to, and a 0 if it doesn't. After the sum, each process knows how many messages to expect from other processes. But even just this one `MPI_Sum` showed up in memory and run time profiles, and so we implemented the algorithm that @gassmoeller abstracted here.)

**Comment by kronbichler** [comment]
I can confirm what was stated by @bangerth - the motivation is to run these algorithms with up to the order of a million MPI ranks. The data communication is O(1) per rank, and we really would like to see this reflected in the implementation, avoiding an `O(P)` implementation. Or to put it differently, the design choice was to design an algorithm that completes in less than a millisecond (1e-3 s) also on 300k MPI ranks. This is reasonable from the point of view that an `Allreduce` of a single element on that scale is around 1e-4 seconds, and 20 or 50 point-to-point messages of a few kB in size, mostly between nearby processes in the MPI universe, also run on this order of time.

I don't think we actually tested `MPI_Alltoallv()` back then, but both `MPI_Allgather` and `MPI_Alltoall` were way too expensive to be usable. I'm talking about seconds instead of milliseconds, with run time getting worse as we increased the computational resources. As you might suspect from this scale, we did not actually use `OpenMPI` on those large experiments but the vendor one, which was Intel MPI in our case. Anyway, the hope with these kinds of algorithms is to be able to use a uniform algorithm on different kinds of machines as we scale through various problem sizes. We might introduce shortcuts for specific configurations, and we actually do so for less than 100 MPI ranks https://github.com/dealii/dealii/blob/9237409a8151cd53c8527575798b9ad5f9aba838/include/deal.II/base/mpi_consensus_algorithms.h#L2470-L2482 - but in general this is our desire.

**Comment by jsquyres** [comment]
After a bunch more conversation about this, and more testing, we found a new issue which we think explains everything.

A key factor was that we were also able to reproduce the issue with the latest version of MPICH.  While we never say "never", the chances of both Open MPI and MPICH having the same bug -- keeping in mind that our internal MPI progression engines are entirely different -- are pretty small.

## Bottom line

The core issue is that there's no protection between successive calls to `run()`.

What can happen is that a "fast" MPI process may exit `run()` before its peers, and then invoke the *next* iteration of `run()`, and therefore call `MPI_Isend(send_buffer, ..., &send_request)` again.   This Isend may get matched in the lower `while(...)` loop (after the Ibarrier) in the processes still in the prior iteration's invocation of `run()`, and may cause a process to call `MPI_Isend(..., &answer_request)` **a second time**.

This orphans a request, which probably leads to a slow memory leak.  It also introduces error in your algorithm, because data is being sent in iteration X, but is being received in iteration X-1.

## Recommendation

The super-simple way to avoid this problem is to put a blocking barrier at the bottom of `run()`.

This probably will add some performance overhead that you're trying to avoid, though, so an alternative still-easy-but-not-trivially-easy is to use different point-to-point tags for each invocation to `run()`.  E.g., use tag set A for even number iterations of `run()` and tag set B for odd number iterations.

**Comment by bangerth** [comment]
@jsquyres  & @bosilca Interesting! Do I read correctly that you managed to reproduce the issue, and that the problem went away when you added the barrier at the end?

@gassmoeller When you have a chance, could you try whether the issue resolves if you put a (regular) barrier at the end of the function? I suspect that the penalty for a barrier is acceptable.

(And if that works, we should instead try to use deal.II's `CollectiveMutex` class. It is currently implemented using simply a regular `MPI_Barrier` call, but we have disabled calls to an `MPI_Ibarrier` that we should test on occasion. In essence, in the `unlock()` function we create an Ibarrier call, and in the `lock()` function we make sure that everyone has gotten out of the protected region. This is essentially a penalty-free way of ensuring that we're never in the same critical code.)

**Comment by bangerth** [comment]
Separately -- I appreciate all of the time being put into this. I know that deal.II is only one of the many users of what you are building here. Thank you for considering our issues so quickly!

**Comment by gassmoeller** [comment]
> @gassmoeller When you have a chance, could you try whether the issue resolves if you put a (regular) barrier at the end of the function? I suspect that the penalty for a barrier is acceptable.
> (And if that works, we should instead try to use deal.II's CollectiveMutex class. 

I will get to it asap next week (I am on a conference this week, little time :-(), but I think we will need to reconsider the solution. At least I am sure that in the production code this algorithm is already inside the `CollectiveMutex` class of deal.II, which enforces a MPI_barrier before the next call of the algorithm (I removed the barrier for the minimal example). I am not sure this enforces that there is no MPI communication **at all** before the barrier though, I will take a look at that.

Again I appreciate all the feedback, sorry to not be more responsive this week.

**Comment by devreal** [comment]
Another solution to the problem could be to use distinct tags in each iteration. That would insulate the sends/recvs/probes of consecutive calls to `run()` without incurring the cost of a barrier.

**Comment by gassmoeller** [comment]
Great, thanks for all the helpful feedback! 
I had some time to look into the issue and do some more tests this week. Let's take this one by one:

1. Thanks for pointing out the additional problem that without a barrier subsequent executions of `run` can cause conflicting messages. I have reproduced your problem with a fresh install of openmpi on my local PC and can confirm that adding a `MPI_Barrier` between calls to `run` fixes the additional problem. I have no idea why this problem does not show up on the cluster that shows the bug I described above. For all further analysis I ran the fixed code (attached below).
2. I have verified that 1. does not fix the problem I described in this issue. We already have an MPI barrier after the algorithm in our production code, and I now introduced a `MPI_Barrier` into the minimal example after each call to `run` and I still observe the bug (infinite loop waiting for completion of `run()`, always after ~2^24 executions of the algorithm).
3. I have compiled my own recent version of openucx (1.13.1) on the affected cluster and compiled openmpi (4.1.4) against it, and now observe the same bug with this fresh installation that I saw with the original setup (openmpi 4.1.1 and ucx 1.9.0 preinstalled by our support). I.e. it is not something that has changed in UCX between 1.9.0 and 1.13.1.
4. Above I wrote that the bug goes away if I use `ob1` instead of `ucx` as the openmpi pml. I have now found that I **can** use UCX successfully if I set UCX's configuration a certain way. Setting the environment variable `UCX_TLS` to either `all` (default) or `ib` (infiniband) will trigger the bug. Setting it to any of `posix`, `shm`, `sysv`, `self`, or `^ib` (not infiniband) does not trigger the bug. It seems only including the infiniband transport layer causes the issue, even though I am running this on a single node (i.e. the infiniband connection should not be needed). I know too little about UCX to know if selecting 'all' should include the infiniband network, even if I am running on a single node. This would explain why you do not see the same issue on your system @jsquyres, it only exists within the infiniband component of UCX (or some configuration or driver for infiniband on our cluster, who knows).

By now it looks to me like this is an issue in UCX (or with my configuration of ucx), not openmpi itself. I also have some good workarounds, although they would not work for multi-node jobs. How should we proceed with this issue?

[ibarrier.cc.txt](https://github.com/open-mpi/ompi/files/10078920/ibarrier.cc.txt)

**Comment by bosilca** [comment]
Let's make sure @open-mpi/ucx, the UCX folks, are well aware of this issue.

**Comment by janjust** [comment]
@brminich seems like an issue in UCX ib TLS'

**Comment by gassmoeller** [comment]
Thanks for all the help. I have no further information on the issue, but after an update of our system the behavior disappeared and everything works as intended, so maybe the issue was fixed already, or it was a configuration problem. I will close the issue for now unless we (or someone else) encounters it again.
